jsonnet/kube-prometheus/kube-state-metrics/kube-state-metrics.libsonnet

```diff
@@ -1,313 +1,53 @@
-      deployment.mixin.spec.template.spec.securityContext.withRunAsUser(65534) +
-      deployment.mixin.spec.template.spec.withServiceAccountName('kube-state-metrics'),
-
-    roleBinding:
-      local roleBinding = k.rbac.v1.roleBinding;
-
-      roleBinding.new() +
-      roleBinding.mixin.metadata.withName('kube-state-metrics') +
-      roleBinding.mixin.metadata.withNamespace($._config.namespace) +
-      roleBinding.mixin.roleRef.withApiGroup('rbac.authorization.k8s.io') +
-      roleBinding.mixin.roleRef.withName('kube-state-metrics') +
-      roleBinding.mixin.roleRef.mixinInstance({ kind: 'Role' }) +
-      roleBinding.withSubjects([{ kind: 'ServiceAccount', name: 'kube-state-metrics' }]),
-
-    role:
-      local role = k.rbac.v1.role;
-      local rulesType = role.rulesType;
-
-      local coreRule = rulesType.new() +
-                       rulesType.withApiGroups(['']) +
-                       rulesType.withResources([
-                         'pods',
-                       ]) +
-                       rulesType.withVerbs(['get']);
-
-      local extensionsRule = rulesType.new() +
-                             rulesType.withApiGroups(['extensions']) +
-                             rulesType.withResources([
-                               'deployments',
-                             ]) +
-                             rulesType.withVerbs(['get', 'update']) +
-                             rulesType.withResourceNames(['kube-state-metrics']);
-
-      local appsRule = rulesType.new() +
-                       rulesType.withApiGroups(['apps']) +
-                       rulesType.withResources([
-                         'deployments',
-                       ]) +
-                       rulesType.withVerbs(['get', 'update']) +
-                       rulesType.withResourceNames(['kube-state-metrics']);
-
-      local rules = [coreRule, extensionsRule, appsRule];
-
-      role.new() +
-      role.mixin.metadata.withName('kube-state-metrics') +
-      role.mixin.metadata.withNamespace($._config.namespace) +
-      role.withRules(rules),
-
-    serviceAccount:
-      local serviceAccount = k.core.v1.serviceAccount;
-
-      serviceAccount.new('kube-state-metrics') +
-      serviceAccount.mixin.metadata.withNamespace($._config.namespace),
-
-    service:
-      local service = k.core.v1.service;
-      local servicePort = service.mixin.spec.portsType;
-
-      local ksmServicePortMain = servicePort.newNamed('https-main', 8443, 'https-main');
-      local ksmServicePortSelf = servicePort.newNamed('https-self', 9443, 'https-self');
-
-      service.new('kube-state-metrics', $.kubeStateMetrics.deployment.spec.selector.matchLabels, [ksmServicePortMain, ksmServicePortSelf]) +
-      service.mixin.metadata.withNamespace($._config.namespace) +
-      service.mixin.metadata.withLabels({ 'k8s-app': 'kube-state-metrics' }) +
-      service.mixin.spec.withClusterIp('None'),
-
+    clusterRole: ksm.clusterRole,
```

Comment:
No need to repeat there you can just specify the servicemonitor where the hidden config fields are specified.

Reply:
not sure, I understood it right. please check the last commit.

Reply:
more like, the whole file can be like this:

```jsonnet
{
  kubeStateMetrics:: (import 'kube-state-metrics/kube-state-metrics.libsonnet') + {
    local ksm = self,
    name:: 'kube-state-metrics',
    namespace:: 'monitoring',
    version:: '1.9.4',
    image:: 'quay.io/coreos/kube-state-metrics:v' + ksm.version,

    serviceMonitor: <service monitor definition using ksm variables>,
  }
}
```

Reply:
fwiw this was probably a mistake: https://github.com/kubernetes/kube-state-metrics/blob/2148cb9bd5ba51d3911a93c4d61ddf084999f8c7/jsonnet/kube-state-metrics/kube-state-metrics.libsonnet#L12

We should remove the "v" prefixing and just use exactly the version string and expect people to set it with the v prefix if they want it. If you want I'd be happy about that upstream contribution :)

Reply:
> more like, the whole file can be like this:

thanks, I will rewrite. 

> fwiw this was probably a mistake: https://github.com/kubernetes/kube-state-metrics/blob/2148cb9bd5ba51d3911a93c4d61ddf084999f8c7/jsonnet/kube-state-metrics/kube-state-metrics.libsonnet#L12

sure.